rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,491 @@
+
+As a start, we propose to add “implements” and “reference” as two attributes
+defining a “standard” function.
+
+For example:
+
+```
+@function.function(
+    implements=“google.matmul_low_rank_matrix”)
+def matmul_lowrank(x, y):
+   “Multiplies two low rank matrices. ”
+   # Basic implementation goes here.
+```
+
+Note we use a wrapper named function, that under-the-hood will create a
+standard tensorflow function with standard attributes set.
+
+Arguments:
+
+* `implements`:  we propose using namespaces with the top-level namespace
+being company name, with a few reserved namespaces indicating built-in ops, 
+if so desired.  This attributes indicates what function this subgraph implements,
+and provides a hint for downstream parsers that they can replace this with 
+their own implementation should they so choose. 
+
+* `reference`:  The goal for this attribute is to provide a stable reference for the function
+implementation details. This could be a reference to non-tensorflow library (e.g. numpy.linalg.special_matmul)
+or arxiv reference. The downstream tooling can use this value as an opaque cache to verify that 
+it implements the right flavor of the function. 
+
+
+This brings us the following advantages:
+
+1) *Backward and fork compatibility*: the tooling by default will,
+ignore `implements` and just rely on function implementation available in the
+graph, thus ensuring that it “just works”. Further, if users have their
+proprietary implementations they can still ship models that would transparently
+work in open source tensorflow.
+
+2) *Easy detection*: Tools can easily detect the functions attribute and substitute with
+its own more efficient implementation if available.
+
+3) *Simplified implementations on custom hardware*: The downstream tooling can
+use provided reference implementation as a source of ground truth (that can run
+on custom hardware out of the box!) when developing custom implementation on
+their hardware.
+
+4) *Reduced implementation dependencies*: Operation maintainer can change the
+implementation without the fear of breaking the downstream tools pattern
+recognition (since there isn’t any.)
+
+5) *Forward compatibility*: Operation maintainer can even add an atomic
+implementation without breaking existing tools. No overhead in keeping the basic
+implementation available.
+
+6) *Simpler automatic conversion* As tensorflow moves forward it is conceivable
+that its computation description language (today: graph def, tomorrow: MLIR) 
+will be fully be separated from the underlying tensorflow kernel implementations.
+This kind of standard attributes allow for more automatic conversions
+should the need arise.
+
+7) Does not change op developer workflow and can be introduced incrementally to
+existing code.
+
+
+### Behind the scenes changes
+When a new tf.function is created, behind the scenes tensorflow creates up-to 3 functions that are stored in the graph. These functions are: 
```

Comment:
Is it true? I thought that @tf.function would not do that, but taking the gradient of a function would create the two extra functions?

Reply:
well the function creation is lazy in the first place... so tf.function creates no functions at all... They are created when you actually invoke them and/or call gradient of them. So i think the statement is technically correct: note "up-to" :)